File: Contracts/V1/ApiRoutes.cs
namespace DragonflyTracker.Contracts.V1
{
    public static class ApiRoutes
    {
        public const string Root = "api";
        public const string Version = "v1";
        public const string Base = Root + "/" + Version;

        public static class Posts
        {
            public const string GetAll = Base + "/posts";
            public const string Update = Base + "/posts/{postId}";
            public const string Delete = Base + "/posts/{postId}";
            public const string Get = Base + "/posts/{postId}";
            public const string Create = Base + "/posts";
        }
        
        public static class Tags
        {
            public const string GetAll = Base + "/tags";            
            public const string Get = Base + "/tags/{tagName}";            
            public const string Create = Base + "/tags";            
            public const string Delete = Base + "/tags/{tagName}";
        }

        public static class Organizations
        {
            public const string GetAll = Base + "/organizations";
            public const string Update = Base + "/organizations/{organizationName}";
            public const string Get = Base + "/organizations/{organizationName}";
            public const string Create = Base + "/organizations";
            public const string Delete = Base + "/organizations/{organizationName}";

            public static class Projects
            {
                public const string GetAll = Organizations.Get + "/projects";
                public const string Update = Organizations.Get + "/projects/{projectName}";
                public const string Get = Organizations.Get + "/projects/{projectName}";
                public const string Create = Organizations.Get + "/projects";
                public const string Delete = Organizations.Get + "/projects/{projectName}";

                public static class ProjectOwners
                {
                    public const string Get = Projects.Get + "/owner";
                    public const string Update = Projects.Get + "/owner";
                }

                public static class ProjectMaintainers
                {
                    public const string Get = Projects.Get + "/maintainers/{adminUserName}";
                    public const string GetAll = Projects.Get + "/maintainers";
                    public const string Create = Projects.Get + "/maintainers";
                    public const string Delete = Projects.Get + "/maintainers/{adminUserName}";
                }

                public static class ProjectAdmins
                {
                    public const string Get = Projects.Get + "/admins/{adminUserName}";
                    public const string GetAll = Projects.Get + "/admins";
                    public const string Create = Projects.Get + "/admins";
                    public const string Delete = Projects.Get + "/admins/{adminUserName}";
                }

                public static class IssueTypes
                {
                    public const string GetAll = Projects.Get + "/issue-types";
                    public const string Update = Projects.Get + "/issue-types/{issueTypeName}";
                    public const string Get = Projects.Get + "/issue-types/{issueTypeName}";
                    public const string Create = Projects.Get + "/issue-types";
                    public const string Delete = Projects.Get  + "/issue-types/{issueTypeName}";
                }

                public static class IssueStages
                {
                    public const string GetAll = Projects.Get + "/issue-stages";
                    public const string Update = Projects.Get + "/issue-stages/{issueStageName}";
                    public const string Get = Projects.Get + "/issue-stages/{issueStageName}";
                    public const string Create = Projects.Get + "/issue-stages";
                    public const string Delete = Projects.Get + "/issue-stages/{issueStageName}";
                }

                public static class Issues
                {
                    public const string GetAll = Projects.Get + "/issues";
                    public const string Update = Projects.Get + "/issues/{issueNumber}";
                    public const string Get = Projects.Get + "/issues/{issueNumber}";
                    public const string Create = Projects.Get + "/issues";
                    public const string Delete = Projects.Get + "/issues/{issueNumber}";

                    public static class IssuePostUpdates
                    {
                        public const string GetAll = Issues.Get + "/issue-post-updates";
                        public const string Update = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                        public const string Get = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                        public const string Create = Issues.Get + "/issue-post-updates";
                        public const string Delete = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                    }

                    public static class IssuePosts
                    {
                        public const string GetAll = Issues.Get + "/issue-posts";
                        public const string Update= Issues.Get + "/issue-posts/{issuePostNumber}";
                        public const string Get = Issues.Get + "/issue-posts/{issuePostNumber}";
                        public const string Create = Issues.Get + "/issue-posts";
                        public const string Delete = Issues.Get + "/issue-posts/{issuePostNumber}";

                        public static class IssuePostReactions
                        {
                            public const string GetAll = IssuePosts.Get + "/issue-posts";
                            public const string Update = IssuePosts.Get + "/issue-posts/{reactionId}";
                            public const string Get = IssuePosts.Get + "/issue-posts/{reactionId}";
                            public const string Create = IssuePosts.Get + "/issue-posts";
                            public const string Delete = IssuePosts.Get + "/issue-posts/{reactionId}";
                        }
                    }

                    public static class IssueTypes
                    {

                        public const string GetAll = Issues.Get + "/issue-types";
                        public const string Update = Issues.Get + "/issue-types/{issueTypeName}";
                        public const string Get = Issues.Get + "/issue-types/{issueTypeName}";
                        public const string Create = Issues.Get + "/issue-types";
                        public const string Delete = Issues.Get + "/issue-types/{issueTypeName}";
                    }

                    public static class IssueStages
                    {

                        public const string Get = Issues.Get+ "/issue-stage";
                        public const string Update = Issues.Get + "/issue-stage";
                    }
                }
            }
        }

        public static class Users
        {
            public const string GetAll = Base + "/users";
            public const string Update = Base + "/users/{username}";
            public const string Get = Base + "/users/{username}";
            public const string Create = Base + "/users";
            public const string Delete = Base + "/users/{username}";

            public static class IssuePosts
            {
                public const string GetAll = Users.Get + "/issue-posts";
            }

            public static class Projects
            {
                public const string GetAll= Users.Get + "/projects";
                public const string Update = Users.Get + "/projects/{projectName}";
                public const string Get = Users.Get + "/projects/{projectName}";
                public const string Create = Users.Get + "/projects";
                public const string Delete = Users.Get + "/projects/{projectName}";

                public static class ProjectOwners
                {
                    public const string Get = Projects.Get + "/owner";
                    public const string Update = Projects.Get + "/owner";
                }

                public static class ProjectMaintainers
                {
                    public const string Get = Projects.Get + "/maintainers/{adminUserName}";
                    public const string GetAll = Projects.Get + "/maintainers";
                    public const string Create = Projects.Get + "/maintainers";
                    public const string Delete = Projects.Get + "/maintainers/{adminUserName}";
                }

                public static class ProjectAdmins
                {
                    public const string Get = Projects.Get + "/admins/{adminUserName}";
                    public const string GetAll = Projects.Get + "/admins";
                    public const string Create = Projects.Get + "/admins";
                    public const string Delete = Projects.Get + "/admins/{adminUserName}";
                }

                public static class IssueTypes
                {
                    public const string GetAll = Projects.Get + "/issue-types";
                    public const string Update = Projects.Get + "/issue-types/{issueTypeName}";
                    public const string Get = Projects.Get + "/issue-types/{issueTypeName}";
                    public const string Create = Projects.Get + "/issue-types";
                    public const string Delete = Projects.Get + "/issue-types/{issueTypeName}";
                }

                public static class IssueStages
                {                                                        
                    public const string GetAll = Projects.Get + "/issue-stages";
                    public const string Update = Projects.Get + "/issue-stages/{issueStageName}";
                    public const string Get = Projects.Get + "/issue-stages/{issueStageName}";
                    public const string Create= Projects.Get + "/issue-stages";
                    public const string Delete = Projects.Get + "/issue-stages/{issueStageName}";

                }

                public static class Issues
                {
                    public const string GetAll = Projects.Get + "/issues";
                    public const string Update = Projects.Get + "/issues/{issueNumber}";
                    public const string Get = Projects.Get + "/issues/{issueNumber}";
                    public const string Create = Projects.Get + "/issues";
                    public const string Delete = Projects.Get + "/issues/{issueNumber}";

                    public static class IssuePostUpdates
                    {
                        public const string GetAll = Issues.Get + "/issue-post-updates";
                        public const string Update = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                        public const string Get = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                        public const string Create = Issues.Get + "/issue-post-updates";
                        public const string Delete = Issues.Get + "/issue-post-updates/{issueUpdateNumber}";
                    }

                    public static class IssuePosts
                    {
                        public const string GetAll = Issues.Get + "/issue-posts";
                        public const string Update = Issues.Get + "/issue-posts/{issuePostNumber}";
                        public const string Get = Issues.Get + "/issue-posts/{issuePostNumber}";
                        public const string Create = Issues.Get + "/issue-posts";
                        public const string Delete = Issues.Get + "/issue-posts/{issuePostNumber}";

                        public static class IssuePostReactions
                        {
                            public const string GetAll = IssuePosts.Get + "/issue-posts";
                            public const string Update = IssuePosts.Get + "/issue-posts/{reactionId}";
                            public const string Get = IssuePosts.Get + "/issue-posts/{reactionId}";
                            public const string Create = IssuePosts.Get + "/issue-posts";
                            public const string Delete = IssuePosts.Get + "/issue-posts/{reactionId}";
                        }
                    }

                    public static class IssueTypes
                    {

                        public const string GetAll = Issues.Get + "/issue-types";
                        public const string Update = Issues.Get + "/issue-types/{issueTypeName}";
                        public const string Get = Issues.Get + "/issue-types/{issueTypeName}";
                        public const string Create = Issues.Get + "/issue-types";
                        public const string Delete = Issues.Get + "/issue-types/{issueTypeName}";
                    }

                    public static class IssueStages
                    {
                        public const string Get = Issues.Get + "/issue-stage";
                        public const string Update = Issues.Get + "/issue-stage";
                    }
                }
            }
        }

        public static class Projects
        {
            public const string GetAllByOrg = Organizations.Get + "/projects";
            public const string UpdateByOrg = Organizations.Get + "/projects/{projectName}";
            public const string GetByOrg = Organizations.Get + "/projects/{projectName}";
            public const string CreateByOrg = Organizations.Get + "/projects";
            public const string DeleteByOrg = Organizations.Get + "/projects/{projectName}";

            public const string GetAllByUser = Users.Get + "/projects";
            public const string UpdateByUser = Users.Get + "/projects/{projectName}";
            public const string GetByUser = Users.Get + "/projects/{projectName}";
            public const string CreateByUser = Users.Get + "/projects";
            public const string DeleteByUser = Users.Get + "/projects/{projectName}";
            public const string GetAll = Base + "/projects-search";
        }

        public static class Issues
        {
            public const string GetAllByOrgProject = Projects.GetByOrg + "/issues";
            public const string UpdateByOrgProject = Projects.GetByOrg + "/issues/{issueNumber}";
            public const string GetByOrgProject = Projects.GetByOrg + "/issues/{issueNumber}";
            public const string CreateByOrgProject = Projects.GetByOrg + "/issues";
            public const string DeleteByOrgProject = Projects.GetByOrg + "/issues/{issueNumber}";

            public const string GetAllByUserProject = Projects.GetByUser + "/issues";
            public const string UpdateByUserProject = Projects.GetByUser + "/issues/{issueNumber}";
            public const string GetByUserProject = Projects.GetByUser + "/issues/{issueNumber}";
            public const string CreateByUserProject = Projects.GetByUser + "/issues";
            public const string DeleteByUserProject = Projects.GetByUser + "/issues/{issueNumber}";
        }

        public static class IssueStages
        {
            public const string GetAllByOrgProject = Projects.GetByOrg + "/issue-stages";
            public const string UpdateByOrgProject = Projects.GetByOrg + "/issue-stages/{issueStageName}";
            public const string GetByOrgProject = Projects.GetByOrg + "/issue-stages/{issueStageName}";
            public const string CreateByOrgProject = Projects.GetByOrg + "/issue-stages";
            public const string DeleteByOrgProject = Projects.GetByOrg + "/issue-stages/{issueStageName}";

            public const string GetAllByUserProject = Projects.GetByUser + "/issue-stages";
            public const string UpdateByUserProject = Projects.GetByUser + "/issue-stages/{issueStageName}";
            public const string GetByUserProject = Projects.GetByUser + "/issue-stages/{issueStageName}";
            public const string CreateByUserProject = Projects.GetByUser + "/issue-stages";
            public const string DeleteByUserProject = Projects.GetByUser + "/issue-stages/{issueStageName}";


            public const string GetByOrgProjectIssue = Issues.GetByOrgProject + "/issue-stage";
            public const string GetByUserProjectIssue = Issues.GetByUserProject + "/issue-stage";
            public const string CreateByOrgProjectIssue = Issues.GetByOrgProject + "/issue-stage";
            public const string CreateByUserProjectIssue = Issues.GetByUserProject + "/issue-stage";
        }

        public static class IssueTypes
        {
            public const string GetAllByOrgProject = Projects.GetByOrg + "/issue-types";
            public const string UpdateByOrgProject = Projects.GetByOrg + "/issue-types/{issueTypeName}";
            public const string GetByOrgProject = Projects.GetByOrg + "/issue-types/{issueTypeName}";
            public const string CreateByOrgProject = Projects.GetByOrg + "/issue-types";
            public const string DeleteByOrgProject = Projects.GetByOrg + "/issue-types/{issueTypeName}";

            public const string GetAllByUserProject = Projects.GetByUser + "/issue-types";
            public const string UpdateByUserProject = Projects.GetByUser + "/issue-types/{issueTypeName}";
            public const string GetByUserProject = Projects.GetByUser + "/issue-types/{issueTypeName}";
            public const string CreateByUserProject = Projects.GetByUser + "/issue-types";
            public const string DeleteByUserProject = Projects.GetByUser + "/issue-types/{issueTypeName}";

            public const string GetAllByOrgProjectIssue = Issues.GetByOrgProject + "/issue-types";
            public const string UpdateByOrgProjectIssue = Issues.GetByOrgProject + "/issue-types/{issueTypeName}";
            public const string GetByOrgProjectIssue = Issues.GetByOrgProject + "/issue-types/{issueTypeName}";
            public const string CreateByOrgProjectIssue = Issues.GetByOrgProject + "/issue-types";
            public const string DeleteByOrgProjectIssue = Issues.GetByOrgProject + "/issue-types/{issueTypeName}";

            public const string GetAllByUserProjectIssue = Issues.GetByUserProject + "/issue-types";
            public const string UpdateByUserProjectIssue = Issues.GetByUserProject + "/issue-types/{issueTypeName}";
            public const string GetByUserProjectIssue = Issues.GetByUserProject + "/issue-types/{issueTypeName}";
            public const string CreateByUserProjectIssue = Issues.GetByUserProject + "/issue-types";
            public const string DeleteByUserProjectIssue = Issues.GetByUserProject + "/issue-types/{issueTypeName}";
        }

        public static class IssuePosts
        {
            public const string GetAllByOrgProjectIssue = Issues.GetByOrgProject + "/issue-posts";
            public const string UpdateByOrgProjectIssue = Issues.GetByOrgProject + "/issue-posts/{issuePostNumber}";
            public const string GetByOrgProjectIssue = Issues.GetByOrgProject + "/issue-posts/{issuePostNumber}";
            public const string CreateByOrgProjectIssue = Issues.GetByOrgProject + "/issue-posts";
            public const string DeleteByOrgProjectIssue = Issues.GetByOrgProject + "/issue-posts/{issuePostNumber}";

            public const string GetAllByUserProjectIssue = Issues.GetByUserProject + "/issue-posts";
            public const string UpdateByUserProjectIssue = Issues.GetByUserProject + "/issue-posts/{issuePostNumber}";
            public const string GetByUserProjectIssue = Issues.GetByUserProject + "/issue-posts/{issuePostNumber}";
            public const string CreateByUserProjectIssue = Issues.GetByUserProject + "/issue-posts";
            public const string DeleteByUserProjectIssue = Issues.GetByUserProject + "/issue-posts/{issuePostNumber}";

            public const string GetAllByUser = Users.Get + "/issue-posts";
        }

        public static class IssuePostReactions
        {
            public const string GetAllByOrg = IssuePosts.GetByOrgProjectIssue + "/issue-posts";
            public const string UpdateByOrg = IssuePosts.GetByOrgProjectIssue + "/issue-posts/{reactionId}";
            public const string GetByOrg = IssuePosts.GetByOrgProjectIssue + "/issue-posts/{reactionId}";
            public const string CreateByOrg = IssuePosts.GetByOrgProjectIssue + "/issue-posts";
            public const string DeleteByOrg = IssuePosts.GetByOrgProjectIssue + "/issue-posts/{reactionId}";

            public const string GetAllByUser = IssuePosts.GetAllByUserProjectIssue + "/issue-posts";
            public const string UpdateByUser = IssuePosts.GetAllByUserProjectIssue + "/issue-posts/{reactionId}";
            public const string GetByUser = IssuePosts.GetAllByUserProjectIssue + "/issue-posts/{reactionId}";
            public const string CreateByUser = IssuePosts.GetAllByUserProjectIssue + "/issue-posts";
            public const string DeleteByUser = IssuePosts.GetAllByUserProjectIssue + "/issue-posts/{reactionId}";
        }

        public static class IssuePostUpdates
        {
            public const string GetAllByOrg = Issues.GetByOrgProject + "/issue-post-updates";
            public const string UpdateByOrg = Issues.GetByOrgProject + "/issue-post-updates/{issueUpdateNumber}";
            public const string GetByOrg = Issues.GetByOrgProject + "/issue-post-updates/{issueUpdateNumber}";
            public const string CreateByOrg = Issues.GetByOrgProject + "/issue-post-updates";
            public const string DeleteByOrg = Issues.GetByOrgProject + "/issue-post-updates/{issueUpdateNumber}";

            public const string GetAllByUser = Issues.GetByUserProject + "/issue-post-updates";
            public const string UpdateByUser = Issues.GetByUserProject + "/issue-post-updates/{issueUpdateNumber}";
            public const string GetByUser = Issues.GetByUserProject + "/issue-post-updates/{issueUpdateNumber}";
            public const string CreateByUser = Issues.GetByUserProject + "/issue-post-updates";
            public const string DeleteByUser = Issues.GetByUserProject + "/issue-post-updates/{issueUpdateNumber}";
        }

        public static class Identity
        {
            public const string Login = Base + "/identity/login";
            public const string Logout = Base + "/identity/logout";
            public const string Register = Base + "/identity/register";
            public const string Refresh = Base + "/identity/refresh";
            public const string Profile = Base + "/identity/profile";
            public const string PasswordReset = Base + "/identity/password-reset";
            public const string PasswordResetEmail = Base + "/identity/password-reset-email";
            public const string PasswordChange = Base + "/identity/password-change";
            public const string EmailConfirm = Base + "/identity/email-confirm";
            public const string EmailConfirmEmail = Base + "/identity/email-confirm-email";
        }

        public static class Notifications
        {
            public const string GetAll = Base + "/notifications";
            public const string Update = Base + "/notifications/{Id}";
            public const string Get = Base + "/notifications/{Id}";
            public const string Create = Base + "/notifications";
            public const string Delete = Base + "/notifications/{Id}";
        }

        public static class AntiForgery {
            public const string Get = Base + "/antiforgery";
        }

        public static class ProjectAdmins {
            public const string GetByUserProject = Projects.GetByUser + "/admins/{adminUserName}";
            public const string GetAllByUserProject = Projects.GetByUser + "/admins";
            public const string CreateByUserProject = Projects.GetByUser + "/admins";
            public const string DeleteByUserProject = Projects.GetByUser + "/admins/{adminUserName}";

            public const string GetByOrgProject = Projects.GetByOrg + "/admins/{adminUserName}";
            public const string GetAllByOrgProject = Projects.GetByOrg + "/admins";
            public const string CreateByOrgProject = Projects.GetByOrg + "/admins";
            public const string DeleteByOrgProject = Projects.GetByOrg + "/admins/{adminUserName}";
        }

        public static class ProjectMaintainers
        {
            public const string GetByUserProject = Projects.GetByUser + "/maintainers/{adminUserName}";
            public const string GetAllByUserProject = Projects.GetByUser + "/maintainers";
            public const string CreateByUserProject = Projects.GetByUser + "/maintainers";
            public const string DeleteByUserProject = Projects.GetByUser + "/maintainers/{adminUserName}";

            public const string GetByOrgProject = Projects.GetByOrg + "/maintainers/{maintainerUserName}";
            public const string GetAllByOrgProject = Projects.GetByOrg + "/maintainers";
            public const string CreateByOrgProject = Projects.GetByOrg + "/maintainers";
            public const string DeleteByOrgProject = Projects.GetByOrg + "/maintainers/{maintainerUserName}";
        }

        public static class ProjectOwners
        {
            public const string GetByUserProject = Projects.GetByUser + "/owner";
            public const string CreateByUserProject = Projects.GetByUser + "/owner";

            public const string GetByOrgProject = Projects.GetByOrg + "/owner";
            public const string CreateByOrgProject = Projects.GetByOrg + "/owner";
        }

    }
}

File: ClientApp/src/app/models/api-routes.ts
import { apiRoot } from '../../environments/environment';

const Base = apiRoot;

const Organizations = {
  GetAll: Base + '/organizations',
  Update: Base + '/organizations/{organizationName}',
  Get: Base + '/organizations/{organizationName}',
  Create: Base + '/organizations',
  Delete: Base + '/organizations/{organizationName}',
};

const Users = {
  GetAll: Base + '/users',
  Update: Base + '/users/{username}',
  Get: Base + '/users/{username}',
  Create: Base + '/users',
  Delete: Base + '/users/{username}',
};

const Projects = {
  GetAllByOrg: Organizations.Get + '/projects',
  UpdateByOrg: Organizations.Get + '/projects/{projectName}',
  GetByOrg: Organizations.Get + '/projects/{projectName}',
  CreateByOrg: Organizations.Get + '/projects',
  DeleteByOrg: Organizations.Get + '/projects/{projectName}',

  GetAllByUser: Users.Get + '/projects',
  UpdateByUser: Users.Get + '/projects/{projectName}',
  GetByUser: Users.Get + '/projects/{projectName}',
  CreateByUser: Users.Get + '/projects',
  DeleteByUser: Users.Get + '/projects/{projectName}',
  GetAll: Base + '/projects-search',
};

const Issues = {
  GetAllByOrgProject: Projects.GetByOrg + '/issues',
  UpdateByOrgProject: Projects.GetByOrg + '/issues/{issueNumber}',
  GetByOrgProject: Projects.GetByOrg + '/issues/{issueNumber}',
  CreateByOrgProject: Projects.GetByOrg + '/issues',
  DeleteByOrgProject: Projects.GetByOrg + '/issues/{issueNumber}',

  GetAllByUserProject: Projects.GetByUser + '/issues',
  UpdateByUserProject: Projects.GetByUser + '/issues/{issueNumber}',
  GetByUserProject: Projects.GetByUser + '/issues/{issueNumber}',
  CreateByUserProject: Projects.GetByUser + '/issues',
  DeleteByUserProject: Projects.GetByUser + '/issues/{issueNumber}',
};

const IssueStages = {
  GetAllByOrgProject: Projects.GetByOrg + '/issue-stages',
  UpdateByOrgProject: Projects.GetByOrg + '/issue-stages/{issueStageName}',
  GetByOrgProject: Projects.GetByOrg + '/issue-stages/{issueStageName}',
  CreateByOrgProject: Projects.GetByOrg + '/issue-stages',
  DeleteByOrgProject: Projects.GetByOrg + '/issue-stages/{issueStageName}',

  GetAllByUserProject: Projects.GetByUser + '/issue-stages',
  UpdateByUserProject: Projects.GetByUser + '/issue-stages/{issueStageName}',
  GetByUserProject: Projects.GetByUser + '/issue-stages/{issueStageName}',
  CreateByUserProject: Projects.GetByUser + '/issue-stages',
  DeleteByUserProject: Projects.GetByUser + '/issue-stages/{issueStageName}',

  GetByOrgProjectIssue: Issues.GetByOrgProject + '/issue-stage',
  GetByUserProjectIssue: Issues.GetByUserProject + '/issue-stage',
  CreateByOrgProjectIssue: Issues.GetByOrgProject + '/issue-stage',
  CreateByUserProjectIssue: Issues.GetByUserProject + '/issue-stage',
};

const IssueTypes = {
  GetAllByOrgProject: Projects.GetByOrg + '/issue-types',
  UpdateByOrgProject: Projects.GetByOrg + '/issue-types/{issueTypeName}',
  GetByOrgProject: Projects.GetByOrg + '/issue-types/{issueTypeName}',
  CreateByOrgProject: Projects.GetByOrg + '/issue-types',
  DeleteByOrgProject: Projects.GetByOrg + '/issue-types/{issueTypeName}',

  GetAllByUserProject: Projects.GetByUser + '/issue-types',
  UpdateByUserProject: Projects.GetByUser + '/issue-types/{issueTypeName}',
  GetByUserProject: Projects.GetByUser + '/issue-types/{issueTypeName}',
  CreateByUserProject: Projects.GetByUser + '/issue-types',
  DeleteByUserProject: Projects.GetByUser + '/issue-types/{issueTypeName}',

  GetAllByOrgProjectIssue: Issues.GetByOrgProject + '/issue-types',
  UpdateByOrgProjectIssue: Issues.GetByOrgProject + '/issue-types/{issueTypeName}',
  GetByOrgProjectIssue: Issues.GetByOrgProject + '/issue-types/{issueTypeName}',
  CreateByOrgProjectIssue: Issues.GetByOrgProject + '/issue-types',
  DeleteByOrgProjectIssue: Issues.GetByOrgProject + '/issue-types/{issueTypeName}',

  GetAllByUserProjectIssue: Issues.GetByUserProject + '/issue-types',
  UpdateByUserProjectIssue: Issues.GetByUserProject + '/issue-types/{issueTypeName}',
  GetByUserProjectIssue: Issues.GetByUserProject + '/issue-types/{issueTypeName}',
  CreateByUserProjectIssue: Issues.GetByUserProject + '/issue-types',
  DeleteByUserProjectIssue: Issues.GetByUserProject + '/issue-types/{issueTypeName}',
};

const IssuePosts = {
  GetAllByOrgProjectIssue: Issues.GetByOrgProject + '/issue-posts',
  UpdateByOrgProjectIssue: Issues.GetByOrgProject + '/issue-posts/{issuePostNumber}',
  GetByOrgProjectIssue: Issues.GetByOrgProject + '/issue-posts/{issuePostNumber}',
  CreateByOrgProjectIssue: Issues.GetByOrgProject + '/issue-posts',
  DeleteByOrgProjectIssue: Issues.GetByOrgProject + '/issue-posts/{issuePostNumber}',

  GetAllByUserProjectIssue: Issues.GetByUserProject + '/issue-posts',
  UpdateByUserProjectIssue: Issues.GetByUserProject + '/issue-posts/{issuePostNumber}',
  GetByUserProjectIssue: Issues.GetByUserProject + '/issue-posts/{issuePostNumber}',
  CreateByUserProjectIssue: Issues.GetByUserProject + '/issue-posts',
  DeleteByUserProjectIssue: Issues.GetByUserProject + '/issue-posts/{issuePostNumber}',

  GetAllByUser: Users.Get + '/issue-posts',
};

const IssuePostReactions = {
  GetAllByOrg: IssuePosts.GetByOrgProjectIssue + '/issue-posts',
  UpdateByOrg: IssuePosts.GetByOrgProjectIssue + '/issue-posts/{reactionId}',
  GetByOrg: IssuePosts.GetByOrgProjectIssue + '/issue-posts/{reactionId}',
  CreateByOrg: IssuePosts.GetByOrgProjectIssue + '/issue-posts',
  DeleteByOrg: IssuePosts.GetByOrgProjectIssue + '/issue-posts/{reactionId}',

  GetAllByUser: IssuePosts.GetAllByUserProjectIssue + '/issue-posts',
  UpdateByUser: IssuePosts.GetAllByUserProjectIssue + '/issue-posts/{reactionId}',
  GetByUser: IssuePosts.GetAllByUserProjectIssue + '/issue-posts/{reactionId}',
  CreateByUser: IssuePosts.GetAllByUserProjectIssue + '/issue-posts',
  DeleteByUser: IssuePosts.GetAllByUserProjectIssue + '/issue-posts/{reactionId}',
};

const IssuePostUpdates = {
  GetAllByOrg: Issues.GetByOrgProject + '/issue-post-updates',
  UpdateByOrg: Issues.GetByOrgProject + '/issue-post-updates/{issueUpdateNumber}',
  GetByOrg: Issues.GetByOrgProject + '/issue-post-updates/{issueUpdateNumber}',
  CreateByOrg: Issues.GetByOrgProject + '/issue-post-updates',
  DeleteByOrg: Issues.GetByOrgProject + '/issue-post-updates/{issueUpdateNumber}',

  GetAllByUser: Issues.GetByUserProject + '/issue-post-updates',
  UpdateByUser: Issues.GetByUserProject + '/issue-post-updates/{issueUpdateNumber}',
  GetByUser: Issues.GetByUserProject + '/issue-post-updates/{issueUpdateNumber}',
  CreateByUser: Issues.GetByUserProject + '/issue-post-updates',
  DeleteByUser: Issues.GetByUserProject + '/issue-post-updates/{issueUpdateNumber}',
};

const Identity = {
  Login: Base + '/identity/login',
  Logout: Base + '/identity/logout',
  Register: Base + '/identity/register',
  Refresh: Base + '/identity/refresh',
  Profile: Base + '/identity/profile',
  PasswordReset: Base + '/identity/password-reset',
  PasswordResetEmail: Base + '/identity/password-reset-email',
  PasswordChange: Base + '/identity/password-change',
  EmailConfirm: Base + '/identity/email-confirm',
  EmailConfirmEmail: Base + '/identity/email-confirm-email',
};

const Notifications = {
  GetAll: Base + '/notifications',
  Update: Base + '/notifications/{Id}',
  Get: Base + '/notifications/{Id}',
  Create: Base + '/notifications',
  Delete: Base + '/notifications/{Id}',
};

const AntiForgery = {
  Get: Base + '/antiforgery',
};

const ProjectAdmins = {
  GetByUserProject: Projects.GetByUser + '/admins/{adminUserName}',
  GetAllByUserProject: Projects.GetByUser + '/admins',
  CreateByUserProject: Projects.GetByUser + '/admins',
  DeleteByUserProject: Projects.GetByUser + '/admins/{adminUserName}',

  GetByOrgProject: Projects.GetByOrg + '/admins/{adminUserName}',
  GetAllByOrgProject: Projects.GetByOrg + '/admins',
  CreateByOrgProject: Projects.GetByOrg + '/admins',
  DeleteByOrgProject: Projects.GetByOrg + '/admins/{adminUserName}',
};

const ProjectMaintainers = {
  GetByUserProject: Projects.GetByUser + '/maintainers/{adminUserName}',
  GetAllByUserProject: Projects.GetByUser + '/maintainers',
  CreateByUserProject: Projects.GetByUser + '/maintainers',
  DeleteByUserProject: Projects.GetByUser + '/maintainers/{adminUserName}',

  GetByOrgProject: Projects.GetByOrg + '/maintainers/{maintainerUserName}',
  GetAllByOrgProject: Projects.GetByOrg + '/maintainers',
  CreateByOrgProject: Projects.GetByOrg + '/maintainers',
  DeleteByOrgProject: Projects.GetByOrg + '/maintainers/{maintainerUserName}',
};

const ProjectOwners = {
  GetByUserProject: Projects.GetByUser + '/owner',
  CreateByUserProject: Projects.GetByUser + '/owner',

  GetByOrgProject: Projects.GetByOrg + '/owner',
  CreateByOrgProject: Projects.GetByOrg + '/owner',
};

export const ApiRoutesV1 = {
  Base,
  Users,
  Organizations,
  Projects,
  Issues,
  IssueStages,
  IssueTypes,
  IssuePosts,
  IssuePostReactions,
  IssuePostUpdates,
  Identity,
  Notifications,
  AntiForgery,
  ProjectAdmins,
  ProjectMaintainers,
  ProjectOwners,
};
